include/alp/structures.hpp

```diff
@@ -802,19 +843,81 @@ namespace alp {
 			using inferred_structures = tuple_cat< std::tuple< Orthogonal >, NonSingular::inferred_structures, OrthogonalColumns::inferred_structures, OrthogonalRows::inferred_structures >::type;
 		};
 
+		template<>
+		struct isInstantiable< RectangularDiagonal, Diagonal > {
+			template< typename ImfR, typename ImfC >
+			static bool check( const ImfR &imf_r, const ImfC &imf_c ) {
+				static_assert( std::is_base_of< imf::Strided, ImfR >::value && std::is_base_of< imf::Strided, ImfC >::value );
+				return ( ( imf_r.n == imf_c.n ) && ( imf_r.b == imf_c.b ) && ( imf_r.s == imf_c.s ) );
+			};
+		};
+
+		template<>
+		struct isInstantiable< General, Diagonal > {
+			template< typename ImfR, typename ImfC >
+			static bool check( const ImfR &imf_r, const ImfC &imf_c ) {
+				static_assert( std::is_base_of< imf::Strided, ImfR >::value && std::is_base_of< imf::Strided, ImfC >::value );
+				return ( ( imf_r.n == imf_c.n ) && ( imf_r.b == imf_c.b ) && ( imf_r.s == imf_c.s ) );
+			};
+		};
+
 		template<>
 		struct isInstantiable< Orthogonal, Orthogonal > {
 			template< typename ImfR, typename ImfC >
 			static bool check( const ImfR &imf_r, const ImfC &imf_c ) {
-				// This check has to be further improved.
-				// Orthogonal matrix in the current implementation
-				// means full-rank square orthogonal matrix.
-				// Rectangular matrix orthogonal only by rows (or columns)
-				// does not fit into the current Orthogonal structure.
 				return imf_r.n == imf_c.n ;
 			};
 		};
 
+		template<>
+		struct isInstantiable< OrthogonalColumns, OrthogonalColumns > {
+			template< typename ImfR, typename ImfC >
+			static bool check( const ImfR &imf_r, const ImfC &imf_c ) {
+				(void) imf_c;
+				return(
+					( imf_r.map( 0 ) == 0 ) &&
```

Comment:
We could check that imf_r is Strided as well: 
```
static_assert( std::is_base_of< imf::Strided, ImfR >::value );
``` 
and that it maps n elements into n elements (this drops the possibility of eliminating rows)